test/MemorySizeTest.cpp

```diff
@@ -0,0 +1,213 @@
+         std::pair<std::string_view, AllMemoryUnitSizes>{
+             "MB", AllMemoryUnitSizes{1000000uL, 1e3, 1, 1e-3, 1e-6}},
+         std::pair<std::string_view, AllMemoryUnitSizes>{
+             "GB", AllMemoryUnitSizes{1000000000uL, 1e6, 1e3, 1, 1e-3}},
+         std::pair<std::string_view, AllMemoryUnitSizes>{
+             "TB", AllMemoryUnitSizes{1000000000000uL, 1e9, 1e6, 1e3, 1}}});
+
+/*
+Checks all the getters of the class with the wanted memory sizes.
+*/
+void checkAllMemorySizeGetter(const ad_utility::MemorySize& m,
+                              const AllMemoryUnitSizes& ms) {
+  ASSERT_EQ(m.bytes(), ms.bytes);
+  ASSERT_DOUBLE_EQ(m.kilobytes(), ms.kilobytes);
+  ASSERT_DOUBLE_EQ(m.megabytes(), ms.megabytes);
+  ASSERT_DOUBLE_EQ(m.gigabytes(), ms.gigabytes);
+  ASSERT_DOUBLE_EQ(m.terabytes(), ms.terabytes);
+}
+
+TEST(MemorySize, MemorySizeConstructor) {
+  // Default constructor.
+  ad_utility::MemorySize m1;
+  checkAllMemorySizeGetter(m1, AllMemoryUnitSizes{0uL, 0.0, 0.0, 0.0, 0.0});
+
+  // Factory functions for size_t overload.
+  checkAllMemorySizeGetter(ad_utility::MemorySize::bytes(1uL),
+                           singleMemoryUnitSizes.at("B"));
+  checkAllMemorySizeGetter(ad_utility::MemorySize::kilobytes(1uL),
+                           singleMemoryUnitSizes.at("kB"));
+  checkAllMemorySizeGetter(ad_utility::MemorySize::megabytes(1uL),
+                           singleMemoryUnitSizes.at("MB"));
+  checkAllMemorySizeGetter(ad_utility::MemorySize::gigabytes(1uL),
+                           singleMemoryUnitSizes.at("GB"));
+  checkAllMemorySizeGetter(ad_utility::MemorySize::terabytes(1uL),
+                           singleMemoryUnitSizes.at("TB"));
+
+  // Factory functions for double overload.
+  checkAllMemorySizeGetter(ad_utility::MemorySize::kilobytes(1.0),
+                           singleMemoryUnitSizes.at("kB"));
+  checkAllMemorySizeGetter(ad_utility::MemorySize::megabytes(1.0),
+                           singleMemoryUnitSizes.at("MB"));
+  checkAllMemorySizeGetter(ad_utility::MemorySize::gigabytes(1.0),
+                           singleMemoryUnitSizes.at("GB"));
+  checkAllMemorySizeGetter(ad_utility::MemorySize::terabytes(1.0),
+                           singleMemoryUnitSizes.at("TB"));
+}
+
+TEST(MemorySize, AssignmentOperator) {
+  ad_utility::MemorySize m;
+  checkAllMemorySizeGetter(m, AllMemoryUnitSizes{0uL, 0.0, 0.0, 0.0, 0.0});
+
+  m = 1_B;
+  checkAllMemorySizeGetter(m, singleMemoryUnitSizes.at("B"));
+
+  m = 1_kB;
+  checkAllMemorySizeGetter(m, singleMemoryUnitSizes.at("kB"));
+
+  m = 1_MB;
+  checkAllMemorySizeGetter(m, singleMemoryUnitSizes.at("MB"));
+
+  m = 1_GB;
+  checkAllMemorySizeGetter(m, singleMemoryUnitSizes.at("GB"));
+
+  m = 1_TB;
+  checkAllMemorySizeGetter(m, singleMemoryUnitSizes.at("TB"));
+}
+
+// For tests, where `MemorySize` is converted into string and vice-versa.
+struct MemorySizeInByteAndStringRepresentation {
+  size_t memoryAmount_;
```

Comment:
I think this member should by of type `MemorySize`, then your code below gets simpler.